cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java

```diff
@@ -379,12 +395,88 @@ private void stopWorkflowAdapter(Id.Namespace namespace, AdapterSpecification ad
+    Id.Application appId = Id.Application.from(namespace, adapterSpec.getTemplate());
+    ApplicationSpecification appSpec = store.getApplication(appId);
+    if (appSpec == null) {
+      throw new NotFoundException(appId);
+    }
+    String workflowName = Iterables.getFirst(appSpec.getWorkers().keySet(), null);
+    return Id.Program.from(namespace.getId(), adapterSpec.getTemplate(), ProgramType.WORKER, workflowName);
+  }
+
+  private void startWorkerAdapter(Id.Namespace namespace, AdapterSpecification adapterSpec) throws NotFoundException {
+    final Id.Adapter adapterId = Id.Adapter.from(namespace.getId(), adapterSpec.getName());
+    final Id.Program workerId = getWorkerId(namespace, adapterSpec);
+    try {
+      Map<String, String> sysArgs = resolver.getSystemProperties(workerId, ProgramType.WORKER);
+      Map<String, String> userArgs = resolver.getUserProperties(workerId, ProgramType.WORKER);
+      // Pass Adapter Name as a system property
+      sysArgs.put(ProgramOptionConstants.ADAPTER_NAME, adapterSpec.getName());
+      // Override resolved preferences with adapter worker spec properties.
+      userArgs.putAll(adapterSpec.getRuntimeArgs());
+      store.setWorkerInstances(workerId, adapterSpec.getInstances());
+      ProgramRuntimeService.RuntimeInfo runtimeInfo = lifecycleService.start(workerId, ProgramType.WORKER,
+                                                                             sysArgs, userArgs, false);
+      final ProgramController controller = runtimeInfo.getController();
+      controller.addListener(new AbstractListener() {
+        @Override
+        public void init(ProgramController.State state, @Nullable Throwable cause) {
+          if (state == ProgramController.State.COMPLETED) {
+            completed();
+          }
+          if (state == ProgramController.State.ERROR) {
+            error(controller.getFailureCause());
+          }
+        }
+
+        @Override
+        public void completed() {
+          super.completed();
+          LOG.debug("Adapter {} completed", adapterId);
+          store.setAdapterStatus(adapterId.getNamespace(), adapterId.getId(), AdapterStatus.STOPPED);
+        }
+
+        @Override
+        public void error(Throwable cause) {
+          super.error(cause);
+          LOG.debug("Adapter {} stopped with error : {}", adapterId, cause);
+          store.setAdapterStatus(adapterId.getNamespace(), adapterId.getId(), AdapterStatus.STOPPED);
+        }
+
+        @Override
+        public void killed() {
+          super.killed();
+          LOG.debug("Adapter {} killed", adapterId);
+          store.setAdapterStatus(adapterId.getNamespace(), adapterId.getId(), AdapterStatus.STOPPED);
+        }
+      }, Threads.SAME_THREAD_EXECUTOR);
+    } catch (Throwable t) {
+      if (t instanceof FileNotFoundException) {
+        throw new NotFoundException(workerId);
+      } else {
+        Throwables.propagate(t);
```

Comment:
should try to avoid doing this whenever possible.  It changes checked exceptions to runtime exceptions means you lose the ability to do proper error handling.  why not include them as exceptions the method can throw?